lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_filter( 'block_core_navigation_render_inner_blocks', 'gutenberg_hook_blocks_into_core_navigation', 10, 3 );
+
+/**
+ * Hooks into the REST API response for the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @param WP_REST_Request  $request  Request object.
+ * @return WP_REST_Response $response The response object.
+ */
+function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $post, $request ) {
+	$content                               = serialize_blocks( gutenberg_hook_first_last_children( $response->data['content']['raw'] ) );
+	$response->data['content']['raw']      = $content;
+	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
+
+	return $response;
+}
+
+/**
+ * Hooks into the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_Block_List  $inner_blocks The inner blocks of the core/navigation block.
+ * @return WP_Block_List $inner_blocks with the first and last inner blocks added.
+ */
+function gutenberg_hook_blocks_into_core_navigation( $inner_blocks ) {
+	return new WP_Block_List( gutenberg_hook_first_last_children( $inner_blocks ) );
+}
+
+/**
+ * Returns the first and last inner blocks of the core/navigation block.
+ *
+ * @param string|WP_Block_List $inner_blocks The inner blocks of the core/navigation block, or post_content string.
+ * @return array parsed inner blocks.
+ */
+function gutenberg_get_parsed_blocks( $blocks ) {
+	if ( is_string( $blocks ) ) {
+		return parse_blocks( $blocks );
+	}
+
+	if ( $blocks instanceof WP_Block_List ) {
+		return array_map(
+			function ( $inner_block ) {
+				return $inner_block->parsed_block;
+			},
+			iterator_to_array( $blocks )
+		);
+	}
+
+	return array();
+}
+
+/**
+ * Adds the first and last inner blocks if there are hooked blocks.
+ *
+ * @param string|WP_Block_List $inner_blocks The inner blocks of the core/navigation block, or post_content string.
+ * @return array $inner_blocks with the first and last inner blocks added.
+ */
+function gutenberg_hook_first_last_children( $inner_blocks ) {
+	$before_block_visitor = null;
+	$after_block_visitor  = null;
+	$hooked_blocks        = get_hooked_blocks();
+	$blocks               = gutenberg_get_parsed_blocks( $inner_blocks );
```

Comment:
TBH I'd rather avoid "type-normalizing" helpers like this; I find that in practice, they tend to make debugging harder.

Let's maybe change `gutenberg_hook_first_last_children` to only accept an array of parsed blocks as its argument (which is also the type it returns, so that's nicely consistent), and get rid of `gutenberg_get_parsed_blocks`; after all, we know whether we're dealing with a serialized string of blocks or with a `WP_Block_List` at the respective callsites of `gutenberg_hook_first_last_children`, so we can massage them into the correct type right there, before passing them to the function.

Reply:
I agree, that was an obvious improvement I missed myself. Updated in https://github.com/WordPress/gutenberg/pull/57754/commits/3092c435f810dd4a2077f2e3c71302ab7cd98275